README
# README --- Information GNU Hyperbole users and maintainers should read
#
# Author:       Bob Weiner
#
# Orig-Date:    19-Oct-91 at 03:27:47
#
# Copyright (C) 1989-2021  Free Software Foundation, Inc.
# See the "HY-COPY" file for license information.
#
# This file is part of GNU Hyperbole.

We hope you enjoy using GNU Hyperbole.  Feel free to mail or post news
containing this file wherever it may be of use.  Hyperbole is designed and
written by Bob Weiner.  It is maintained by him and Mats Lidell.

===========================================================================
*			Table of Contents
===========================================================================
			* Summary
			* Files
			* Programmer Quick Reference
			* User Quotes
			* Why was Hyperbole developed?

===========================================================================
*			   Summary
===========================================================================

GNU Hyperbole (pronounced Ga-new Hi-per-bo-lee), or just Hyperbole, is like
Markdown for hypertext.  Hyperbole automatically recognizes dozens of
common, pre-existing patterns in any buffer regardless of mode and can
instantly activate them as hyperbuttons with a single key: email addresses,
URLs, grep -n outputs, programming backtraces, sequences of Emacs keys,
programming identifiers, Texinfo and Info cross-references, Org links,
Markdown links and on and on.  All you do is load Hyperbole and then your
text comes to life with no extra effort or complex formatting.

Hyperbole includes easy-to-use, powerful hypertextual button types without
the need to learn a markup language.  Hyperbole's button types are written
in Lisp and can be wholly independent of the web, i.e. web links are one
type of Hyperbole link, not fundamental to its link architecture.  However,
Hyperbole is a great assistant when editing HTML or Javascript or when
browsing web pages and links.

Hyperbole comes pre-built with most of the implicit button types you will
need but with a little extra effort and a few lines of code (or even just a
few words), you can define your own implicit button types to recognize your
specific buttons and then activate them anywhere in Emacs.  You press a
single key, {M-RET} by default, on any kind of Hyperbole button to activate
it, so you can rely on your muscle memory and let the computer do the hard
work of figuring out what to do.  {C-u M-RET} shows you what any button will
do in any context before you activate it, so you can always be sure of what
you are doing when needed or if some emails you a button (you can do that
too).

Hyperbole is something to be experienced and interacted with, not understood
from reading alone.  It installs normally as a single Emacs package with no
dependencies outside of standard Emacs libraries.  Most of Hyperbole is a
single global minor mode that you can activate and deactivate at will.  And
it can be uninstalled quickly as well if need be, so there is no risk to
giving it a spin.

Once you have it installed and activated {C-u M-x hyperbole-mode RET}, try
the interactive demo with {C-h h d d}.  In fact, if you have Hyperbole
loaded, you can press {M-RET} inside any of the brace delimited series of
keys you see in this document and it will execute them on-the-fly (easy
keyboard-macro style buttons in any text).

Hyperbole can dramatically increase your productivity and greatly reduce the
number of keyboard/mouse keys you'll need to work efficiently.

In short, Hyperbole lets you:

1. Quickly create hyperlink buttons either from the keyboard or by dragging
between a source and destination window with a mouse button depressed.
Later, activate buttons by pressing/clicking on them or by giving the name of
the button.

2. Activate many kinds of `implicit buttons' recognized by context within
text buffers, e.g. URLs, grep output lines, and git commits.  A single key
or mouse button automatically does the right thing in dozens of contexts;
just press and go.

3. Build outlines with multi-level numbered outline nodes, e.g. 1.4.8.6,
that all renumber automatically as any node or tree is moved in the
outline. Each node also has a permanent hyperlink anchor that you can
reference from any other node;

4. Manage all your contacts quickly with hierarchical categories and embed
hyperlinks within each entry. Or create an archive of documents with
hierarchical entries and use the same search mechanism to quickly find any
matching entry;

5. Use single keys to easily manage your Emacs windows or frames and quickly
retrieve saved window and frame configurations;

6. Search for things in your current buffers, in a directory tree or across
major web search engines with the touch of a few keys.

The common thread in all these features is making retrieval, management and
display of information fast and easy. That is Hyperbole's purpose.

Hyperbole works well on GNU Emacs 24.4 or above. It is designed and written
by Bob Weiner.  It is maintained by him and Mats Lidell.  Its main
distribution site is: https://www.gnu.org/software/hyperbole.  If any term in
here is new or unfamiliar to you, you can look it up in the Hyperbole Manual
Glossary.

===========================================================================
*			   Files
===========================================================================

For a more extensive summary of Hyperbole with text and images, see the
Hyperbole home page: https://www.gnu.org/software/hyperbole.  If you have
already downloaded Hyperbole, see the "HY-ABOUT" file for a longer
description and overview of Hyperbole.

See "DEMO" for a demonstration of standard Hyperbole button capabilities.
This is the best way to initially interactively learn about Hyperbole after
installing it.

"man/im/demo.png", "man/im/hyperbole-cv.png" and "koutline.png" are
screenshots of Hyperbole in action.

HyControl is Hyperbole's frame and window manager; a long video
demonstrating many of HyControl's features is available at
https://youtu.be/M3-aMh1ccJk.

See the "HY-NEWS" file for a summary of new features in this release.

See the "INSTALL" file for installation and invocation instructions.

See the "HY-COPY" and "COPYING" files for license information.

See the "MANIFEST" file for summaries of Hyperbole distribution files.

Various forms of the reference manual for Hyperbole are below the
"man/" subdirectory.

===========================================================================
*		     Programmer Quick Reference
===========================================================================

"MANIFEST" summarizes most of the files in the distribution.

See "DEMO" for a demonstration of standard Hyperbole button capabilities.
This is the best way to initially interactively learn about Hyperbole.
The Hyperbole Manual is a reference manual, not a simple introduction.

Naming conventions:

  - All Hyperbole-specific code files begin with an 'h', aside from the
    Koutliner files which are in the kotl/ subdirectory and begin with a 'k'.

  - Hyperbole user-interface files begin with 'hui-' or 'hmous'.

  - Files that define implicit button types begin with 'hib'.

  - Encapsulations of foreign systems begin with 'hsys-'.

Most of the standard Emacs user interface for Hyperbole is located in
"hui.el".  Most of the Hyperbole application programming interface can be
found in "hbut.el".  "hbdata.el" encapsulates the button attribute storage
implemented by Hyperbole.  "hmail.el" provides a basic abstract interface
for integrating mail readers other than Rmail into Hyperbole.

See the "(hyperbole)Questions and Answers" appendix in the Hyperbole
manual for information on how to alter the default context-sensitive
Hyperbole key bindings (Smart Keys).

===========================================================================
*			    User Quotes
===========================================================================


  *** MAN I love Hyperbole!!!  Wow! ***

				    	-- Ken Olstad
					   Cheyenne Software, Inc.

-------

  I *love* koutlines.

	   				-- Bob Glickstein
					   Z-Code Software Corporation

-------

  One of the nicest things about Hyperbole is that it's available
  everywhere. Org-mode is a mode and its features are only available in
  *.org files. For instance if you dropped into `eshell' or `ansi-term' and
  did `ls', you can move point to any of the directory's contents, do M-RET
  (or Shift-Button2) and jump to that file.  And that's just one example.
  Note that this means that all Hyperbole functionality is available in
  *.org files as well.  To me, except for the Hyperbole outliner, that means
  complementary not conflicting. It's Hyperbole *and* org-mode, not
  Hyperbole vs. org-mode.

  Additionally, off the bat, I found it very well documented and for me
  that's a proxy for the quality of a package.  The maintainers are quite
  responsive.  There's plenty more functionality that I haven't uncovered yet
  but due to the ease of installation and the quality of the documentation,
  digging into it is actually fun.

					-- Aditya Siram
-------

  For me, Emacs isn't Emacs without Hyperbole.  I have depended on Hyperbole
  daily since 1992, when I first started using it to manage my development
  environment.  It didn't take long before I could summon almost any
  information I needed directly from within my editing environment with an
  implicit button. Since I almost never have to slow down to look for
  things--one context-dependent button usually produces exactly what I need
  --I am able to maintain focus on the task I am working on and complete it
  more quickly.  With its gestural interface, seamless integration with other
  Emacs packages and incredibly useful set of core features.  I think that
  Hyperbole is one of the best designed and most easily extensible software
  products I have ever come across.  It is certainly the one which has made
  the biggest improvement in my personal productivity.

					-- Chris Nuzum
					   Co-founder, Traction Software, Inc.

-------

  I've found Hyperbole (in conjunction with XEmacs) to be very useful
  for signal processing algorithm development.

  For me, it has almost completely obsoleted the engineering notebook:
  I keep a set of files with ideas, algorithms, and results, linked
  together and to the implementation in C++ files.  Using XEmacs'
  support for embedding graphics, I've written a mode that accepts
  image tags (formatted like HTML), and reads in GIF files to display
  plots.  I have another program that converts the file to HTML (not
  perfect, but adequate), so I can put any aspect of development on
  our internal web for others to see.

                                        -- Farzin Guilak
		                           Protocol Systems, Inc., Engineer

-------

  I am blind and have been using Hyperbole since 1992.  I used to use a PC as
  a talking terminal attached to a UNIX system, but then I developed
  Emacspeak which lets me use Emacs and Hyperbole from standard UNIX
  workstations with an attached voice synthesizer.

  My main uses are:
    1) Global and implicit buttons for jumping to ftp sites.
    2) The contact manager with Emacspeak support.
    3) Explicit buttons as part of comments made about a structured document.
       Each button jumps to the document section referred to by the comment.
       This is very, very useful.
    4) The Hyperbole Koutliner, which I find a very useful tool.  I've
       implemented Emacspeak extensions to support it.

					-- TV Raman
					   Google Inc.

-------

  I've been a grateful Hyperbole user for a few years now.  Hyperbole's
  flexibility and ease of use is a marvel.

  Mainly, I write easy little implicit button types (and corresponding action
  types) to make my life easier.  For example, I have an implicit button type
  to bury certain buffers when I click at their bottoms, one that recognizes
  a bug report record in various contexts and edits it, one that links pieces
  of test output in a log file to the corresponding test case source code
  (EXTREMELY helpful in interpreting test output), others that support our
  homegrown test framework, one that handles tree dired mode the way I'd
  like, one that completely handles wico menus (I've also overloaded the
  wconfig actions triggered by diagonal mouse drags with wicos actions), and
  a couple that support interaction with BBDB.

  Other than that, I keep a global button file with 30 or so explicit buttons
  that do various little things, and I index saved mail messages by putting
  explicit link-to-mail buttons in an outline file.

				    	-- Ken Olstad
					   Cheyenne Software, Inc.

-------

  In general, Hyperbole is an embeddable, highly extensible hypertext
  tool.  As such, I find it very useful. As it stands now, Hyperbole is
  particularly helpful for organizing ill-structured or loosely coupled
  information, in part because there are few tools geared for this purpose.
  Hyperbole also possesses a lot of potential in supporting a wider
  spectrum of structuredness, ranging from unstructured to highly
  structured environments, as well as structural changes over time.

  Major Uses:

  * Menu interface to our own collaborative support environment called
    CoReView: This interface brings together all top-level user commands
    into a single partitioned screen, and allows the end user to interact
    with the system using simple mouse-clicking instead of the meta-x key.

  * Gateway to internet resources: this includes links to major Internet
    archive sites of various types of information. Links are made at both
    directory and file levels.

  * Alternative directory organizer: The hierarchical nature of the Unix
    file system sometimes makes it difficult to find things quickly and
    easily using directory navigational tools such as dired. Hyperbole
    enables me to create various "profile" views of my directory tree, with

Changes
2021-05-01  Bob Weiner  <rsw@gnu.org>

* hypb-ert.el (hypb-ert-sym, hypb-ert-sel): Change 2nd args to
    strings to fix Emacs 27 byte-compile error.

2021-04-30  Bob Weiner  <rsw@gnu.org>

* hpath.el (hpath:at-p):
  hibtypes.el (pathname): Prevent from matching to a shell
    command within a semicolon separated sequence of commands.

2021-04-29  Bob Weiner  <rsw@gnu.org>

* hyperbole.el (hyperb:init): Remove enhanced priority Hyperbole
    minor-mode from emulation-mode-map-alists.
               (hkey-initialize): Remove Hyperbole-specific C-h
    and C-c prefix maps and use of the corresponding global prefix
    maps as parent keymaps for hyperbole-mode.  define-key does
    the right thing for minor modes without any of this.
               (hkey-help-set-key): Remove, not needed.

2021-04-28  Mats Lidell  <matsl@gnu.org>

* Makefile (EL_COMPILE) (ELC_COMPILE): Add hui-dired-sidebar and
    hypb-maintenance

2021-04-28  Bob Weiner  <rsw@gnu.org>

* hyrolo.el (hyrolo-grep-or-fgrep): Add.
            (hyrolo-mode-map): Repurpose {r} to do a new regexp
    HyRolo search, {C-u r} for string search, using above
    new command.  This key had moved to previous matches but
    that is already bound to {Shift-TAB} and {M-TAB}.
            (hyrolo-bbdb-grep, hyrolo-google-contacts-grep): Let
    current-prefix-arg be nil after captured as arg so does not
    affect hyrolo-f/grep calls.
  man/hyperbole.texi (HyRolo Keys): Document new {r} binding.

* hyperbole.el (hyperbole-mode): Add hyperbole-mode-hook.
               (hyperbole--enable-mode, hyperbole--disable-mode):
    Add to separate from hyperb:init.
               (hyperbole-mode): Update to use above functions.

* hui-mini.el (hui:hypb-exit): "X" to exit from hyperbole-mode
    when a minibuffer menu is active.  Use {C-h h} to re-enable
    hyperbole-mode and display the minibuffer menu.
              (hui:hypb-exit): Command invoked by X in menus.

2021-04-27  Bob Weiner  <rsw@gnu.org>

* hui-mouse.el (smart-magit-display-file): Add and use in smart-magit*
    functions so when files from diffs are displayed, they utilize
    Hyperbole's referent display setting, hpath:display-where.
  man/hyperbole.texi (Smart Key - Magit Mode): Updated Smart Magit
    doc with the above.

* hbut.el (gbut:ebut-program): Add to programmatically add global
    explicit buttons at the end of the personal button file.

* hmouse-sh.el (hmouse-get-bindings, hmouse-get-unshifted-bindings):
  hmouse-key.el (hmouse-check-action-key, hmouse-check-assist-key,
    hmouse-set-bindings): Change from global bindings to hyperbole-mode
    ones.

* hui-select.el (hui-select-at-p): Change test from '(and (bobp) (eobp))'
    to '(eobp)' only so never triggers at the end of a buffer, notably
    Hyperbole minibuffer menu help displays.

* hyperbole.el (hyperbole-mode-specific-map): Define to shadow without
    overwriting global C-c bindings.

2021-04-27  Mats Lidell  <matsl@gnu.org>

* test/hypb-tests.el (hypb:replace-match-string-after-27.1-test): Use
    separate test for replace matches that were fixed after Emacs 27.1.

* test/hui-tests.el (hui-ibut-label-create-fails-if-label-exists): Use
    string-match.

* test/hbut-tests.el (hbut-ib-create-label-fails-if-label-exists)
(hbut-pathname-dot-slash-in-other-folder-should-fail-test): Use string-match.

2021-04-26  Bob Weiner  <rsw@gnu.org>

* hui.el (hui:global-bind-key): Renamed from hui:bind-key and made that
    name bind keys within the hyperbole minor mode map.

* hui-em-but.el (hproperty:but-add): Add make-overlay 4th arg = t to prevent
    highlighting from spreading to text inserted before an ebut.

* hbut.el (hbut:key-to-label): Ensure is a string and if one char, is not
    a punctuation char.
          (ibut:to): Fix possible infinite loop if label being tested is one
    one character long.  This caused Action Key on "." in path variable values
    to hang.

* hpath.el (hpath:path-variable-value-regexp, hpath:at-p): Improved regexp
    structure and expanded classes of excluded chars to prevent spurious path
    matches by hpath:delimited-possible-path.

* hui-mini.el (hyperbole): Enable hyperbole-mode when invoked with {C-h h}.

* hyperbole.el (hyperbole-help-map, hkey-help-set-key): Add to set help keys
    available only when Hyperbole is enabled.

2021-04-26  Mats Lidell  <matsl@gnu.org>

* test/hib-org-tests.el (hib-org:org-mode-with-smart-keys-on-file-is-hypb-button)
    (hib-org:org-mode-with-smart-keys-buttons-on-file-is-hypb-button)
    (hib-org:org-mode-with-no-smart-keys-on-file-is-hypb-button): Tests for
    hsys-org-enable-smart-keys.

* test/hy-test-helpers.el
    (hy-test-helpers:hypb-function-should-call-hpath:find): Test helper to
    verify any hypb function that should call hpath:find

2021-04-25  Mats Lidell  <matsl@gnu.org>

* test/hsys-org-tests.el (hsys-org:cycle-on-header-cycles-visibility)
    (hsys-org:region-with-text-property-value, hsys-org:block-start-at-p)
    (hsys-org:org-link-at-p, hsys-org:org-target-at-p)
    (hsys-org:org-radio-target-link-at-p)
    (hsys-org:org-radio-target-def-at-p, hsys-org:org-radio-target-at-p)
    (hsys-org:org-internal-link-target-at-p, hsys-org:org-face-at-p): hsys-org tests

* test/hib-org-tests.el (hib-org:org-mode-on-header-cycles-visibility):
    hib-org test

* hsys-org.el (hsys-org-link-at-p): Use org-target-link-regexp for radio links
    (hsys-org-internal-link-target-at-p): Inverse boolean check

2021-04-25  Bob Weiner  <rsw@gnu.org>

* hyperbole.el (hkey-override-local-bindings,
                hkey-install-override-local-bindings,
                hkey-init-override-local-keys,
                hkey-toggle-override-local-bindings): Remove, replaced
    by minor mode.
  man/hyperbole.texi (Hyperbole Key Bindings): Rename from Global
    Key Bindings.
                     (Default Hyperbole Bindings): Rename from
    Default Global Bindings and remove discussion of overriding keys.

htz.el
;;; htz.el --- Timezone-based time and date support for GNU Hyperbole
;;
;; Author:       Masanobu Umeda             / Bob Weiner
;;
;; Orig-Date:    14-Oct-91 at 07:22:08
;;
;; Copyright (C) 1991-2016  Free Software Foundation, Inc.
;; See the "HY-COPY" file for license information.
;;
;; Derived from "timezone.el" in GNU Emacs.
;;
;; This file is part of GNU Hyperbole.
;;
;;; Commentary:
;;
;; All date parsing functions accept the output of any other parsing
;; function as input, so one can convert to a sortable date format, do a
;; compare and the display the result in a user selected format.
;; All date formats use a 4-digit year, so there are no problems around the
;; turn of the century.
;;
;; Hyperbole uses this package to normalize all worldwide date times to
;; Universal Time (Greenwich Mean Time), so that Hyperbole buttons
;; created by users in different timezones will sort by time properly.

;;; Code:
;;; ************************************************************************
;;; Other required Elisp libraries
;;; ************************************************************************

(require 'calendar)
(require 'cal-julian)

;;; ************************************************************************
;;; Public functions
;;; ************************************************************************

(defun htz:date-arpa (&optional date local timezone)
  "Convert optional DATE or current date to an arpanet standard date.
Optional 1st argument LOCAL specifies the default local timezone of the DATE.
Optional 2nd argument TIMEZONE specifies a timezone to be represented in."
  (or (vectorp date)
      (setq date (htz:date-parse (or date (current-time-string)))))
  (let* ((year   (string-to-number (aref date 0)))
	 (month  (string-to-number (aref date 1)))
	 (day    (string-to-number (aref date 2)))
	 (time   (htz:time-parse   (aref date 3)))
	 (hour   (string-to-number (aref time 0)))
	 (minute (string-to-number (aref time 1)))
	 (second (string-to-number (aref time 2)))
	 (local  (or (aref date 4) local htz:local)) ;Use original if defined
	 (timezone (or timezone local))
	 (diff   (- (htz:zone-to-hour timezone)
		    (htz:zone-to-hour local)))
	 (new    (htz:time-fix year month day
			       (+ hour diff) minute second)))
    (htz:date-make-arpa (aref new 0) (aref new 1) (aref new 2)
			(htz:time-make-string
			 (aref new 3) (aref new 4) (aref new 5))
			timezone)))

(defun htz:date-parse (date &optional parsed-current-date)
  "Parse DATE string and return a vector [year month day time timezone].
If a two-digit year, the first two digits of the current year are prepended.
Timezone in DATE is optional, it defaults to the value of `htz:local'.

Recognizes the following styles:
 (1) \"(1 30 1999)\" or \"(1 30 1999)\"  `calendar-julian-date'  requires PARSED-CURRENT-DATE arg
 (2) \"14 Apr 89 03:20[:12] [GMT]\"
 (3) \"Fri, 17 Mar [19]89 4:01[:33] [GMT]\"
 (4) \"Mon Jan 16 16:12[:37] [GMT] 1989\"
 (5) \"19911014:07:51:08 or 1991101407:51:08\"  `sortable date'
 (6) \"Mar 29 14:00\"    `ls -l date'  requires PARSED-CURRENT-DATE arg
 (7) \"Mar  7  1994\"    `ls -l date'  requires PARSED-CURRENT-DATE arg"
  (let ((date (or date ""))
	year month day time
	zone)			; This may be nil.
    (if (listp date)
	(setq month (nth 0 date)
	      day   (nth 1 date)
	      year  (nth 2 date))
      (cond ((string-match
	      "\\`(\\([0-9]+\\) \\([0-9]+\\) \\([0-9]+\\))\\'" date)
	     ;; Style (1)
	     (setq year 3 month 1 day 2 time nil zone nil))
	    ((string-match
	      ;; Allow for 3 digits in hour to handle prior error in
	      ;; generating hours fixed on 2019-06-10; 3rd digit
	      ;; removed in htz:time-make-string and htz:time-parse.
	      "\\([0-9][0-9][0-9][0-9]\\)\\([0-1][0-9]\\)\\([0-3][0-9]\\):?\\([0-9][0-9][0-9]?:[0-5][0-9:]+\\)[ ]*\\'" date)
	     ;; Style (5)
	     (setq year 1 month 2 day 3 time 4 zone nil))
	    ((string-match
	      "\\([0-9]+\\) \\([^ ,]+\\) \\([0-9]+\\) \\([0-9]+:[0-9:]+\\)[ ]*\\'" date)
	     ;; Styles: (2) and (3) without timezone
	     (setq year 3 month 2 day 1 time 4 zone nil))
	    ((string-match
	      "\\([0-9]+\\) \\([^ ,]+\\) \\([0-9]+\\) \\([0-9]+:[0-9:]+\\)[ ]*\\([-+a-zA-Z0-9]+\\)" date)
	     ;; Styles: (2) and (3) with timezone and buggy timezone
	     (setq year 3 month 2 day 1 time 4 zone 5))
	    ((string-match
	      "\\([^ ,]+\\) +\\([0-9]+\\) \\([0-9]+:[0-9:]+\\(:[0-9]+\\)?\\) \\([0-9]+\\)" date)
	     ;; Styles: (4) without timezone
	     (setq year 5 month 1 day 2 time 3 zone nil))
	    ((string-match
	      "\\([^ ,]+\\) +\\([0-9]+\\) \\([0-9]+:[0-9:]+\\(:[0-9]+\\)?\\) \\([-+a-zA-Z0-9]+\\) \\([0-9]+\\)" date)
	     ;; Styles: (4) with timezone
	     (setq year 6 month 1 day 2 time 3 zone 5))
	    ((string-match "^\\([^ ,]+\\) +\\([0-9]+\\) +\\([0-9]+:[0-9:]+\\)$" date)
	     ;; Style: (6)
	     (setq year nil month 1 day 2 time 3 zone nil))
	    ((string-match
	      "^\\([^ ,]+\\) +\\([0-9]+\\) +\\([0-9][0-9][0-9][0-9]\\)$" date)
	     ;; Style: (7)
	     (setq year 3 month 1 day 2 time nil zone nil))
	    (t (error "(htz:date-parse): Invalid date format: `%s'" date)))
      (if year
	  (setq year
		(substring date (match-beginning year) (match-end year))
		year (if (/= (length year) 2) year
		       (let* ((yr (substring (current-time-string) -4))
			      (curr-yr (substring yr 2))
			      (century (substring yr 0 2)))
			 (concat (if (string< curr-yr yr)
				     (format "%02d"
					     (1- (string-to-number century)))
				   century)
				 year))))
	(setq year (if (vectorp parsed-current-date)
		       (aref parsed-current-date 0)
		     "0")))
      (if month
	  (setq month (substring date
				 (match-beginning month) (match-end month))
		month (if (/= (string-to-number month) 0) month
			(int-to-string
			 (cdr (assoc (upcase month) htz:months-assoc)))))
	(setq month (if (vectorp parsed-current-date)
			(aref parsed-current-date 1)
		      "0")))
      (if day
	  (setq day (substring date (match-beginning day) (match-end day)))
	(setq day (if (vectorp parsed-current-date)
		      (aref parsed-current-date 2)
		    "0"))))
    (if time
	(setq time (substring date (match-beginning time) (match-end time)))
      (setq time (if (vectorp parsed-current-date)
		     (aref parsed-current-date 3))))
    (if zone
	(setq zone (substring date (match-beginning zone) (match-end zone)))
      (setq zone (if (vectorp parsed-current-date)
		     (aref parsed-current-date 4)
		   htz:local)))
    ;; Return a vector.
    (vector year month day time zone)))

(defun htz:date-sortable (&optional date local timezone)
  "Convert optional DATE or current date to a sortable date string.
Optional 1st argument LOCAL specifies the local timezone of the DATE.
Optional 2nd argument TIMEZONE specifies an output timezone to use."
  (or (vectorp date)
      (setq date (htz:date-parse (or date (current-time-string)))))
  (let* ((year   (string-to-number (aref date 0)))
	 (month  (string-to-number (aref date 1)))
	 (day    (string-to-number (aref date 2)))
	 (time   (htz:time-parse   (aref date 3)))
	 (hour   (string-to-number (aref time 0)))
	 (minute (string-to-number (aref time 1)))
	 (second (string-to-number (aref time 2)))
	 (local  (or (aref date 4) local htz:local)) ;Use original if defined
	 (timezone (or timezone local))
	 (diff   (- (htz:zone-to-hour timezone)
		    (htz:zone-to-hour local)))
	 (new    (htz:time-fix year month day
			       (+ hour diff) minute second)))
    (htz:date-make-sortable (aref new 0) (aref new 1) (aref new 2)
			    (htz:time-make-string
			     (aref new 3) (aref new 4) (aref new 5)))))

;;;
;;; Parsers and Constructors of Date and Time
;;;

(defun htz:date-sortable-gmt (&optional date local)
  "Convert optional DATE or current date  to a sortable date string in Greenwich Mean Time.
Optional argument LOCAL specifies the local timezone of the DATE."
  (htz:date-sortable date local "GMT"))

(defun htz:date-unix (&optional date local timezone)
  "Convert DATE or current date to a unix standard date.
Optional 1st argument LOCAL specifies the local timezone of the DATE (default
is the timezone embedded in the date or if there is none, then the value of
`htz:local').  Optional 2nd argument TIMEZONE specifies the timezone in which
the date is returned; it defaults to the value of `htz:local'."
  (or (vectorp date)
      (setq date (htz:date-parse (or date (current-time-string)))))
  (or local (setq local (or (aref date 4) htz:local)))
  (let* ((year   (string-to-number (aref date 0)))
	 (month  (string-to-number (aref date 1)))
	 (day    (string-to-number (aref date 2)))
	 (time   (htz:time-parse (aref date 3)))
	 (hour   (string-to-number (aref time 0)))
	 (minute (string-to-number (aref time 1)))
	 (second (string-to-number (aref time 2)))
	 (timezone (or timezone local))
	 (diff   (- (htz:zone-to-hour timezone)
		    (htz:zone-to-hour local)))
	 (fixed    (htz:time-fix year month day
				 (+ hour diff) minute second)))
    (htz:date-make-unix
     (aref fixed 0) (aref fixed 1) (aref fixed 2)
     (htz:time-make-string (aref fixed 3) (aref fixed 4) (aref fixed 5))
     timezone)))

(defun htz:span-in-days (start-date end-date)
  "Return span in days between START-DATE and END-DATE strings.
See `htz:date-parse' for a list of acceptable date formats."
  (if (and (listp start-date) (listp end-date))
      (- (calendar-julian-to-absolute end-date)
	 (calendar-julian-to-absolute start-date))
    (let* ((parsed-current-date (htz:date-parse (current-time-string)))
	   (htz-start-date (htz:date-parse start-date parsed-current-date))
	   (htz-end-date   (htz:date-parse end-date parsed-current-date))
	   (cal-start-date
	    (list (string-to-number (aref htz-start-date 1));; month
		  (string-to-number (aref htz-start-date 2));; day
		  (string-to-number (aref htz-start-date 0))));; year
	   (cal-end-date
	    (list (string-to-number (aref htz-end-date 1));; month
		  (string-to-number (aref htz-end-date 2));; day
		  (string-to-number (aref htz-end-date 0))));; year
	   )
      (- (calendar-julian-to-absolute cal-end-date)
	 (calendar-julian-to-absolute cal-start-date)))))

(defun htz:time-parse (time)
  "Parse TIME (HH:MM:SS) and return a vector [hour minute second]."
  (let ((time (or time ""))
	(hour nil)
	(minute nil)
	(second nil))
    (cond ((string-match "\\`\\([0-9]+\\):\\([0-9]+\\):\\([0-9]+\\)\\'" time)
	   ;; HH:MM:SS
	   (setq hour 1 minute 2 second 3))
	  ((string-match "\\`\\([0-9]+\\):\\([0-9]+\\)\\'" time)
	   ;; HH:MM
	   (setq hour 1 minute 2 second nil)))
    ;; Return [hour minute second]
    (vector
     (if hour
	 ;; Remove possible 3rd digit in hour to handle prior error in
	 ;; generating hours fixed on 2019-06-10; 3rd digit
	 ;; removed in htz:time-make-string and here.
	 (format "%02.2s" (substring time (match-beginning hour) (match-end hour))) "0")
     (if minute
	 (substring time (match-beginning minute) (match-end minute)) "0")
     (if second
	 (substring time (match-beginning second) (match-end second)) "0"))))

;;; ************************************************************************
;;; Private functions
;;; ************************************************************************

(defun htz:date-make-arpa (year month day time &optional timezone)
  "Make arpanet standard date string from YEAR, MONTH, DAY, and TIME.
Optional argument TIMEZONE specifies a time zone."
  (format "%02d %s %02d %s%s"
	  day
	  (capitalize (car (rassq month htz:months-assoc)))
	  (- year (* (/ year 100) 100))	;1990 -> 90
	  time
	  (if timezone (concat " " timezone) "")
	  ))

(defun htz:date-make-unix (year month day time &optional timezone)
  "Approximate Unix date format from YEAR, MONTH, DAY, and TIME.
Optional argument TIMEZONE specifies a time zone."
  (format "%s %02d %s%s %04d"
	  (capitalize (car (rassq month htz:months-assoc)))
	  day time (if timezone (concat " " timezone) "") year))

(defun htz:date-make-sortable (year month day time)
  "Make sortable date string from YEAR, MONTH, DAY, and TIME."
  (format "%04d%02d%02d:%s" year month day time))

(defun htz:last-day-of-month (month year)
  "The last day in MONTH during YEAR."
  (if (and (= month 2) (htz:leap-year-p year))
      29
    (aref [31 28 31 30 31 30 31 31 30 31 30 31] (1- month))))

(defun htz:leap-year-p (year)
  "Return t if YEAR is a Gregorian leap year."
  (or (and (zerop  (% year 4))
	   (not (zerop (% year 100))))
      (zerop (% year 400))))

(defun htz:time-fix (year month day hour minute second)
  "Fix date and time."
  (cond ((<= 24 hour)			; 24 -> 00
	 (setq hour (- hour 24))
	 (setq day  (1+ day))
	 (if (< (htz:last-day-of-month month year) day)
	     (progn
	       (setq month (1+ month))
	       (setq day 1)
	       (if (< 12 month)
		   (progn
		     (setq month 1)
		     (setq year (1+ year)))))))
	((> 0 hour)
	 (setq hour (+ hour 24))
	 (setq day  (1- day))
	 (if (> 1 day)
	     (progn
	       (setq month (1- month))
	       (if (> 1 month)
		   (progn
		     (setq month 12)
		     (setq year (1- year))))
	       (setq day (htz:last-day-of-month month year))))))
  (vector year month day hour minute second))

;; Partly copied from Calendar program by Edward M. Reingold.
(defun htz:time-make-string (hour minute second)
  "Make time string from HOUR, MINUTE, and SECOND."
  (format "%02.2d:%02.2d:%02.2d" hour minute second))

(defun htz:zone-to-hour (timezone)
  "Translate TIMEZONE (in zone name or integer) to integer hour."
  (if timezone
      (progn
	(setq timezone
	      (or (cdr (assoc (upcase timezone)
			      htz:world-timezones))
		  (and (fboundp 'current-time-zone)
		       (if (listp (current-time-zone))
			   (/ (car (current-time-zone)) 36)
			 (current-time-zone)))
		  timezone))
	(if (stringp timezone)
	    (setq timezone (string-to-number timezone)))
	(/ timezone 100))
    (error "(htz:zone-to-hour): Nil timezone sent as argument")))


;;; ************************************************************************
;;; Private variables
;;; ************************************************************************

(defvar htz:local
  (let ((local-tz (or (getenv "TZ") (getenv "TIMEZONE")
		      (if (fboundp 'current-time-zone)
			  (car (cdr (current-time-zone))))
		      (progn
			(require 'hypb)
			(hypb:call-process-p
			 "date" nil '(if (re-search-backward
					  " \\([-+a-zA-Z0-9]+\\) [0-9]+$" nil t)
					 (buffer-substring (match-beginning 1)
							   (match-end 1))))))))
    (if (and (stringp local-tz) (string-match " " local-tz))
	;; Windows returns things like "Eastern Daylight Time", so
	;; abbreviate to the first letter of each word.
	(concat (mapcar (lambda (s) (aref s 0)) (split-string local-tz)))
      local-tz))
  "Holds string giving the timezone for the local machine.")

(defvar htz:world-timezones
  '(("PST" .  -800)
    ("PDT" .  -700)
    ("MST" .  -700)
    ("MDT" .  -600)
    ("CST" .  -600)
    ("CDT" .  -500)
    ("EST" .  -500)
    ("EDT" .  -400)
    ("AST" .  -400)
    ("NST" .  -330)
    ("UT"  .  +000)
    ("GMT" .  +000)
    ("BST" .  +100)
    ("MET" .  +100)
    ("EET" .  +200)
    ("JST" .  +900)
    ("GMT+1"  .  +100) ("GMT+2"  .  +200) ("GMT+3"  .  +300)
    ("GMT+4"  .  +400) ("GMT+5"  .  +500) ("GMT+6"  .  +600)
    ("GMT+7"  .  +700) ("GMT+8"  .  +800) ("GMT+9"  .  +900)
    ("GMT+10" . +1000) ("GMT+11" . +1100) ("GMT+12" . +1200) ("GMT+13" . +1300)
    ("GMT-1"  .  -100) ("GMT-2"  .  -200) ("GMT-3"  .  -300)
    ("GMT-4"  .  -400) ("GMT-5"  .  -500) ("GMT-6"  .  -600)
    ("GMT-7"  .  -700) ("GMT-8"  .  -800) ("GMT-9"  .  -900)
    ("GMT-10" . -1000) ("GMT-11" . -1100) ("GMT-12" . -1200))
  "Time differentials of timezone from GMT in +-HHMM form.
Use `current-time-zone' instead where possible.")

(defvar htz:months-assoc
  '(("JAN" .  1)("FEB" .  2)("MAR" .  3)
    ("APR" .  4)("MAY" .  5)("JUN" .  6)
    ("JUL" .  7)("AUG" .  8)("SEP" .  9)
    ("OCT" . 10)("NOV" . 11)("DEC" . 12))
  "Alist of first three letters of a month and its numerical representation.")


(provide 'htz)

;;; htz.el ends here
